File: demos/csharp-feequote/src/FeeQuote/Core/Pricing.cs
namespace FeeQuote.Core;

/// <summary>
/// Limits applied to a fee after rounding. The fee is raised to
/// <see cref="MinimumCents"/> and lowered to <see cref="CapCents"/>.
/// </summary>
public sealed record FeePolicy
{
    public FeePolicy(long minimumCents, long capCents)
    {
        if (minimumCents < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(minimumCents), "Minimum fee must not be negative.");
        }

        if (capCents < minimumCents)
        {
            throw new ArgumentOutOfRangeException(nameof(capCents), "Cap must not be below the minimum fee.");
        }

        MinimumCents = minimumCents;
        CapCents = capCents;
    }

    public long MinimumCents { get; }

    public long CapCents { get; }
}

/// <summary>
/// Fee arithmetic shared by every fee type. All inputs and outputs are whole
/// cents; the only division happens in <see cref="RoundHalfEvenDiv"/> so the
/// rounding rule is applied in exactly one place.
/// </summary>
public static class Pricing
{
    /// <summary>Basis points in one whole unit (100 percent).</summary>
    public const int BasisPointsPerUnit = 10_000;

    /// <summary>Rate for the domestic transfer fee.</summary>
    public const int DomesticRateBps = 90;

    /// <summary>Minimum and cap for the domestic transfer fee.</summary>
    public static FeePolicy DomesticPolicy { get; } = new(minimumCents: 100, capCents: 2_500);

    /// <summary>
    /// Divides <paramref name="numerator"/> by <paramref name="denominator"/> and
    /// rounds the quotient half to even (banker's rounding). Both inputs are
    /// integers so the midpoint test is exact.
    /// </summary>
    public static long RoundHalfEvenDiv(long numerator, long denominator)
    {
        if (numerator < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(numerator), "Numerator must not be negative.");
        }

        if (denominator <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(denominator), "Denominator must be positive.");
        }

        long quotient = numerator / denominator;
        long twiceRemainder = (numerator % denominator) * 2;

        if (twiceRemainder > denominator)
        {
            return quotient + 1;
        }

        if (twiceRemainder < denominator)
        {
            return quotient;
        }

        // Exactly halfway: round to the even neighbour.
        return quotient % 2 == 0 ? quotient : quotient + 1;
    }

    /// <summary>
    /// Prices an amount at <paramref name="rateBps"/> basis points, rounds half
    /// to even, then applies the policy minimum and cap in that order.
    /// </summary>
    public static long PriceWithPolicy(long amountCents, int rateBps, FeePolicy policy)
    {
        ArgumentNullException.ThrowIfNull(policy);

        if (amountCents < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(amountCents), "Amount must not be negative.");
        }

        if (rateBps < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(rateBps), "Rate must not be negative.");
        }

        long rounded = RoundHalfEvenDiv(checked(amountCents * rateBps), BasisPointsPerUnit);
        return Math.Clamp(rounded, policy.MinimumCents, policy.CapCents);
    }

    /// <summary>The domestic transfer fee for an amount.</summary>
    public static long ComputeFee(long amountCents)
    {
        return PriceWithPolicy(amountCents, DomesticRateBps, DomesticPolicy);
    }
}

File: demos/typescript-feequote/src/core/pricing.ts
// Core pricing. A fee is a rate in basis points applied to an amount in integer
// cents, rounded half to even, then held within a FeePolicy (minimum and cap).
//
// Amounts cross the public API as safe integers because that is what the rest
// of the codebase and JSON handle naturally. The division itself runs in bigint
// so no intermediate value is ever a binary float.

export interface FeePolicy {
  readonly minimumCents: number;
  readonly capCents: number;
}

export const DOMESTIC_RATE_BPS = 90;

export const DOMESTIC_POLICY: FeePolicy = Object.freeze({
  minimumCents: 100,
  capCents: 2500,
});

const BPS_PER_WHOLE = 10000n;

function assertCents(value: number, name: string): void {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(`${name} must be a non-negative integer number of cents, got ${value}`);
  }
}

/**
 * Integer division that rounds the exact quotient half to even (banker's
 * rounding). Ties go to the even neighbour, so 1111.5 becomes 1112 and 1112.5
 * also becomes 1112.
 */
export function roundHalfEvenDiv(numerator: bigint, denominator: bigint): bigint {
  if (denominator === 0n) {
    throw new RangeError('denominator must not be zero');
  }
  const negative = numerator < 0n !== denominator < 0n;
  const n = numerator < 0n ? -numerator : numerator;
  const d = denominator < 0n ? -denominator : denominator;

  let quotient = n / d;
  const twiceRemainder = (n % d) * 2n;
  if (twiceRemainder > d || (twiceRemainder === d && quotient % 2n === 1n)) {
    quotient += 1n;
  }
  return negative ? -quotient : quotient;
}

/**
 * Prices an amount at a rate in basis points under a policy. The raw fee is
 * rounded first; the minimum and the cap are applied to the rounded value.
 * Every fee type in FeeQuote is built on this function.
 */
export function priceWithPolicy(amountCents: number, rateBps: number, policy: FeePolicy): number {
  assertCents(amountCents, 'amountCents');
  if (!Number.isSafeInteger(rateBps) || rateBps < 0) {
    throw new RangeError(`rateBps must be a non-negative integer, got ${rateBps}`);
  }
  if (policy.minimumCents > policy.capCents) {
    throw new RangeError('policy minimum must not exceed its cap');
  }

  const rounded = Number(roundHalfEvenDiv(BigInt(amountCents) * BigInt(rateBps), BPS_PER_WHOLE));
  if (rounded < policy.minimumCents) {
    return policy.minimumCents;
  }
  if (rounded > policy.capCents) {
    return policy.capCents;
  }
  return rounded;
}

/** The domestic transfer fee. */
export function computeFee(amountCents: number): number {
  return priceWithPolicy(amountCents, DOMESTIC_RATE_BPS, DOMESTIC_POLICY);
}
